lightgbm/src/main/scala/com/microsoft/azure/synapse/ml/lightgbm/TaskTrainingMethods.scala

```diff
@@ -14,16 +14,16 @@ object TaskTrainingMethods {
     * Otherwise, returns true for all tasks.
     * @param trainParams The training parameters.
     * @param log The logger.
+    * @param sharedState The shared state.
     * @return Whether the current task is enabled.
     */
   def isWorkerEnabled(trainParams: TrainParams, log: Logger, sharedState: SharedState): Boolean = {
     if (trainParams.executionParams.useSingleDatasetMode) {
+      // Note: the first valid worker with non-empty partitions sets
+      // the main executor worker, other workers read it
+      sharedState.linkMainExecutorWorker()
```

Comment:
IMHO, this is creating an undocumented side effect in what is basically a getter, and relies on the fact that this method is not called from LightGBMBase if the partition is empty. A more robust fix would call linkMainExecutionWorker in the trainLightGBM method before this is called, specifically conditioned on the partition being empty.
I realize that mainExecutionWorker needs to be initialized for this method to be called, but you could just throw if it's not.

Reply:
good idea!  will update

Reply:
done!